### DISCUSSION TITLE
Expose retrieveSubscriberPackageVersionId() from PackageVersionCreate class

### DISCUSSION THREAD

**Discussion by ClayChipps** [forcedotcom/cli#2721] [Ideas]
Currently, the @salesforce/packaging [PackageVersionCreate class ](https://github.com/forcedotcom/packaging/blob/8b6fe351d693b5799aa507807bf0039a67d54311/src/package/packageVersionCreate.ts) has a function [retrieveSubscriberPackageVersionId](https://github.com/forcedotcom/packaging/blob/8b6fe351d693b5799aa507807bf0039a67d54311/src/package/packageVersionCreate.ts#L150), however it is currently a private function.

There are a few CLI plugins out there that perform dependency installation based on the contents of a sfdx-project.json. Many of these plugins use a similar function for "resolving" the combination Package2Id + VersionNumber to a SubscriberPackageVersionId. It would be extremely beneficial to be able to use this implementation instead of writing our own.

Is this something that we could either expose publicly, or move into to the [packageUtils class](https://github.com/forcedotcom/packaging/blob/8b6fe351d693b5799aa507807bf0039a67d54311/src/utils/packageUtils.ts)?

**Comment by shetzel** [comment]
@ClayChipps - Moving to packageUtils and exposing publicly seems reasonable.   You're more than welcome to open a PR on that repo if you'd like, otherwise I'll add this to our queue for prioritization.

**Reply by ClayChipps** [reply]
Thanks @shetzel! I am working on a couple of things over on the oclif side, but once I wrap up a few things over there I am happy to work on the PR for this. Just wanted to make sure you were good with the approach before doing the work. 👍

**Comment by VivekMChawla** [comment]
@ClayChipps - I spent some time with @shetzel this morning discussing this idea.  

Would a utility function like `resolveSubscriberPackageVersionId(package2Id:string, versionNumber:string):string` that returns the `04t` ID matching the provided Package2Id + VersionNumber combo provide what you need?

Is there additional context around your uses cases that you'd like to share?

Thanks!

**Reply by ClayChipps** [reply]
Hey @VivekMChawla! Appreciate you following up on this. Let me clarify a little bit on what I was looking to do.

There are a lot of really useful functions within the [packageVersionCreate.ts](https://github.com/forcedotcom/packaging/blob/8b6fe351d693b5799aa507807bf0039a67d54311/src/package/packageVersionCreate.ts) class that could be super useful for folks when building packaging plugins.

1. resolveBuildNumber - This would be really helpful for plugins that are trying to manage upgrading dependencies in a sfdx-project.json
2. retrieveSubscriberPackageVersionId - This was the one I originally was looking for, which would be helpful when I am trying to install dependencies from a sfdx-project.json into an org
3. getAncestorId - Not something I would personally use, but again potentially a useful function for someone building a packaging plugin

All of these functions are actually pretty good as is! I actually like that functions like `retrieveSubscriberPackageVersionId` takes an object as the params rather than named parameters. My reason being is, the example signature you provided actually doesn't meet [my needs](https://github.com/SimplySF/simply-package/blob/e950dd31313f1cd35daaeb037de82bfeafbebd5c/src/common/packageUtils.ts#L46) as I also need to consider a potential `branch` value when resolving the version. I would rather keep the params as an object so that we can be flexible in our ability to add additional "criteria" for SubscriberPackageVersion Id resolution in the future without breaking backwards compatibility. 

My biggest issue with the existing functions is simply (other than the fact they are private) that I have no way to specify the `Connection` object that I want to use. All of the functions use `this.connection` which is not ideal for usage as general purpose api.

As I mentioned to @shetzel originally, I would be happy to contribute to this. I really don't have an immediate need for this as I had already developed the same general logic directly in my plugin. I had been working to adopt more of the apis provided by the "base" cli libraries when I ran into this and figured it would be a good opportunity to provide the functionality that was already there to cli developers. :)

**Reply by shetzel** [reply]
@ClayChipps - does [this PR](https://github.com/forcedotcom/packaging/pull/563) address your usecases?

**Reply by ClayChipps** [reply]
@shetzel This is exactly what I was looking for!! Thank you all so much, this just significantly reduced the footprint of my plugin. 🥇
